File: src/descriptors.cpp
#include "descriptors.hpp"

/* ---------------------------------------------------------------------------------------
 */
/* Description: detects landmarks in a left oriented cropped ear image */
/* Parameters: */
/*     Mat img: 96x96 image with a cropped ear */
/*              - left oriented */
/*              - with or without pose variation */
/*     vector<Point2d> &ldmk: vector that receives the 2d coordinates of 55
 * landmarks      */
/*     stage s: FIRST locates landmarks in ears with large pose variations
 */
/*              SECOND locates landmarks in ears with coarse pose
 * normalization
 */
/* Return: none */
/* ---------------------------------------------------------------------------------------
 */
void _detectLandmarks(Mat img, vector<Point2d> &ldmk)
{
    static bool first = true;
    static Net net;

    if (first)
    {
        first = false;
        net = readNetFromTensorflow("model-stage1.pb");
        if (net.empty())
        {
            cerr << "ERROR: Could not load the CNNs for landmark detection"
                 << endl
                 << flush;
            exit(1);
        }
    }
    Mat result, inputBlob = blobFromImage(img);
    inputBlob /= 255.0;
    net.setInput(inputBlob);
    result = net.forward("ear_ang45_3_sca20_r_tra20_r_e/out/MatMul");
    result *= 48;
    result += 48;

    ldmk.clear();
    for (int i = 0; i < 55; i++)
        ldmk.push_back(Point2d(result.at<float>(0, i * 2),
                               result.at<float>(0, i * 2 + 1)));
}

void detectLandmark(vector<vector<Mat>> processedROI,
                    vector<vector<vector<Point2d>>> &landmarks,
                    vector<string> imageNames)
{
    for (int i = 0; i < processedROI.size(); i++)
    {
        printProgress(i, processedROI.size());

        auto images = processedROI[i];
        auto imageName = imageNames[i];

        vector<vector<Point2d>> ldmks;
        for (int j = 0; j < images.size(); j++)
        {
            auto imageROI = images[j];
            vector<Point2d> ldmk;
            _detectLandmarks(imageROI, ldmk);
            ldmks.push_back(ldmk);

            Mat outImage;
            drawLandmarks(imageROI, ldmk, outImage);
            // Exporting landmark image
            String id = to_string(i) + "_" + to_string(j);
            writeToFile(imageName, LANDMARK_PATH, outImage, id);
        }
        landmarks.push_back(ldmks);
    }
}

void reduceDataSparsity(vector<Point2d> points, vector<Point2d> &outPoints,
                        int k)
{
    Point2d centroid(computeCentroid(points));
    // vector<Point2d> _points(points);
    // points.clear();
    outPoints.clear();

    vector<double> distances;
    double mean, std;
    for (auto p : points)
        distances.push_back(norm(p - centroid));
    computeMeanAndStd(distances, mean, std);

    for (int i = 0; i < points.size(); i++)
    {
        if (distances[i] < k * std)
            outPoints.push_back(points[i]);
    }
    if (outPoints.size() < 2)
        outPoints = points;
}

void extractFeatures(vector<vector<Mat>> images,
                     vector<vector<vector<Point2d>>> &landmarks,
                     vector<string> imageNames)
{
    vector<vector<vector<KeyPoint>>> keypoints;
    vector<vector<Mat>> _;
    extractFeatures(images, keypoints, _);

    // Translate KeyPoint -> Point2d
    for (auto image : keypoints)
    {
        vector<vector<Point2d>> imgLdmks;
        for (auto kpts : image)
        {
            vector<Point2d> ldmks;
            for (auto k : kpts)
            {
                ldmks.push_back(Point2d((double)k.pt.x, (double)k.pt.y));
            }
            imgLdmks.push_back(ldmks);
        }
        landmarks.push_back(imgLdmks);
    }

    vector<vector<vector<Point2d>>> outLandmarks;
    for (auto image : landmarks)
    {
        vector<vector<Point2d>> outLdmks;
        for (auto ldmk : image)
        {
            vector<Point2d> outLdmk;
            reduceDataSparsity(ldmk, outLdmk, 3);
            outLdmks.push_back(outLdmk);
        }
        outLandmarks.push_back(outLdmks);
    }

    for (int i = 0; i < images.size(); i++)
    {
        auto image = images[i];
        auto imageName = imageNames[i];
        for (int j = 0; j < image.size(); j++)
        {
            Mat outImage;
            drawLandmarks(image[j], landmarks[i][j], outImage);
            drawLandmarks(outImage, outLandmarks[i][j], outImage,
                          Scalar(0, 255, 0), 5);
            landmarks[i][j] = outLandmarks[i][j];
            // Exporting landmark image
            String id = to_string(i) + "_" + to_string(j);
            writeToFile(imageName, LANDMARK_PATH, outImage, id);
        }
    }
}

void exportFeatures(vector<Mat> descriptors, vector<string> imageNames,
                    const string path)
{
    FileStorage fs(path, FileStorage::WRITE);
    fs << FEATURES_KEY << descriptors << NAMES_KEY << imageNames;
    fs.release();
}

void importFeatures(vector<Mat> &descriptors, vector<string> &imageNames,
                    const string path)
{
    FileStorage fs(path, FileStorage::READ);
    fs[FEATURES_KEY] >> descriptors;
    fs[NAMES_KEY] >> imageNames;
    fs.release();
}

void extractFeatures(vector<vector<Mat>> images, vector<Mat> &descriptors,
                     vector<string> &imageNames, int edgeThreshold,
                     InputArray mask)
{
    vector<vector<vector<KeyPoint>>> keypoints;
    vector<vector<Mat>> tmpDescriptors;
    extractFeatures(images, keypoints, tmpDescriptors);

    vector<string> outNames(imageNames);
    imageNames.clear();
    descriptors.clear();

    // Flatten the descriptor matrix to a 1D vector, where each
    // element corresponds 1:1 to an element of vector imageNames
    for (int i = 0; i < tmpDescriptors.size(); i++)
    {
        int idx = outNames[i].rfind(".");
        for (auto descriptor : tmpDescriptors[i])
        {
            descriptors.push_back(descriptor);
            imageNames.push_back(outNames[i].substr(0, idx));
        }
    }
}

void extractFeatures(vector<vector<Mat>> images,
                     vector<vector<vector<KeyPoint>>> &keypoints,
                     vector<vector<Mat>> &descriptors, int edgeThreshold,
                     InputArray mask)
{
    Ptr<FeatureDetector> detector = ORB::create(500, 1.2f, 8, edgeThreshold);

    for (int i = 0; i < images.size(); i++)
    {
        printProgress(i, images.size());
        auto image = images[i];
        vector<Mat> imageDescriptors;
        vector<vector<KeyPoint>> imageKeypoints;

        for (auto roi : image)
        {
            vector<KeyPoint> ROIkeypoints;
            Mat roiDescriptors;

            detector->detectAndCompute(roi, mask, ROIkeypoints, roiDescriptors);
            imageDescriptors.push_back(roiDescriptors);
            imageKeypoints.push_back(ROIkeypoints);
        }
        descriptors.push_back(imageDescriptors);
        keypoints.push_back(imageKeypoints);
    }
}

float computeSimilarity(Mat queryDescriptors, Mat objectDescriptors,
                        int normType, float ratio, bool crossCheck)
{
    vector<vector<DMatch>> matches;
    Ptr<DescriptorMatcher> matcher = BFMatcher::create(normType, crossCheck);
    matcher->knnMatch(queryDescriptors, objectDescriptors, matches, 2);

    vector<DMatch> goodMatches;
    for (auto match : matches)
    {
        if (match[0].distance < ratio * match[1].distance)
        {
            goodMatches.push_back(match[0]);
        }
    }
    return (float)goodMatches.size() / (float)matches.size();
}

void logSimilarities(Mat queryDescriptor, vector<Mat> imageDescriptors,
                     String queryName, vector<string> imageNames,
                     bool filterByPrefix)
{
    cout << "\nSimilarities with " << queryName << endl;
    vector<string> outNames;
    vector<float> outScores;
    for (int i = 0; i < imageDescriptors.size(); i++)
    {
        auto objectDescriptor = imageDescriptors[i];
        auto imageName = imageNames[i];
        float score = computeSimilarity(queryDescriptor, objectDescriptor);

        string queryPrefix = queryName.substr(0, 3);
        if (!filterByPrefix || startsWith(imageName, queryPrefix))
        {
            outScores.push_back(score);
            outNames.push_back(imageName);
        }
    }
    for (auto i : argSort(outScores, false))
    {
        printf("%.6f - %s\n", outScores[i], outNames[i].c_str());
    }
}

double calculateVerificationGAR(vector<Mat> imageDescriptors, vector<string> imageNames,
                                double threshold)
{
    // Since we only attempt with the given identity templates, this corresponds to the
    // number of true acceptances and false rejections (so false acceptances and true
    // rejections are not considered).
    double attemptsNo = 0.0;
    double acceptancesNo = 0.0;

    // Number of templates per identity
    std::map<std::string, int> templatesNoPerIdentity;

    for (int i = 0; i < imageNames.size(); i++)
    {
        string identityPrefix = imageNames[i].substr(0, 3);

        if (templatesNoPerIdentity.find(identityPrefix) == templatesNoPerIdentity.end())
        {
            templatesNoPerIdentity[identityPrefix] = 0;
        }
        templatesNoPerIdentity[identityPrefix] += 1;

        bool accepted = false;
        for (int j = 0; j < imageNames.size(); j++)
        {
            // Avoiding self comparison
            if (i == j)
            {
                continue;
            }
            if (!startsWith(imageNames[j], identityPrefix))
            {
                continue;
            }
            if (1 - computeSimilarity(imageDescriptors[i], imageDescriptors[j]) <= threshold)
            {
                accepted = true;
            }
        }
        if (accepted)
        {
            acceptancesNo++;
        }
    }

    for (auto x : templatesNoPerIdentity)
    {
        if (x.second > 1)
        {
            attemptsNo += x.second;
        }
    }

    //cout << acceptancesNo << "\n";
    //cout << attemptsNo << "\n";

    return acceptancesNo / attemptsNo;
}

double calculateVerificationFAR(vector<Mat> imageDescriptors, vector<string> imageNames,
                                double threshold)
{
    // Since we only attempt with templates belonging to identities different from
    //  the given one, this corresponds to the number of false acceptances and true
    //  rejections (so true acceptances and false rejections are not considered).
    double attemptsNo = 0.0;
    double acceptancesNo = 0.0;

    for (int i = 0; i < imageNames.size(); i++)
    {
        // map for this probe: identity -> acceptance status
        // initially it is not accepted by any identity
        // when one identity accepts it, it turns to true
        std::map<std::string, bool> acceptedForIdentity;

        string identityPrefix = imageNames[i].substr(0, 3);
        for (int j = 0; j < imageNames.size(); j++)
        {
            if (startsWith(imageNames[j], identityPrefix))
            {
                continue;
            }
            string templateIdentityPrefix = imageNames[j].substr(0, 3);
            // Initializing map entry for this identity
            if (acceptedForIdentity.find(templateIdentityPrefix) == acceptedForIdentity.end())
            {
                acceptedForIdentity[templateIdentityPrefix] = false;
                attemptsNo++;
            }
            if (1.0 - computeSimilarity(imageDescriptors[i], imageDescriptors[j]) <= threshold)
            {
                if (!acceptedForIdentity[templateIdentityPrefix])
                {
                    acceptedForIdentity[templateIdentityPrefix] = true;
                    acceptancesNo++;
                }
            }
        }
    }
    //cout << acceptancesNo << "\n";
    //cout << attemptsNo << "\n";

    return acceptancesNo / attemptsNo;
}

File: android/app/src/main/java/com/getchlabs/earrecognizer/recognition/descriptors.kt
package com.getchlabs.earrecognizer.recognition

import org.opencv.core.*
import org.opencv.core.Core.NORM_HAMMING
import org.opencv.features2d.BFMatcher
import org.opencv.features2d.DescriptorMatcher
import org.opencv.features2d.ORB.create
import kotlin.math.pow
import kotlin.math.sqrt


/* ---------------------------------------------------------------------------------------
 */
/* Description: detects landmarks in a left oriented cropped ear image */
/* Parameters: */
/*     Mat img: 96x96 image with a cropped ear */
/*              - left oriented */
/*              - with or without pose variation */
/*     vector<Point2d> &ldmk: vector that receives the 2d coordinates of 55
 * landmarks      */
/*     stage s: FIRST locates landmarks in ears with large pose variations
 */
/*              SECOND locates landmarks in ears with coarse pose
 * normalization
 */
/* Return: none */
/* ---------------------------------------------------------------------------------------
 */
/*
fun _detectLandmarks(img: Mat, ldmk: ArrayList<Point>)
{
    var first = true;
    var net = Net()

    if (first)
    {
        first = false;
        net = readNetFromTensorflow("model-stage1.pb");
        if (net.empty())
        {
            cerr << "ERROR: Could not load the CNNs for landmark detection"
            << endl
            << flush;
            exit(1);
        }
    }
    Mat result, inputBlob = blobFromImage(img);
    inputBlob /= 255.0;
    net.setInput(inputBlob);
    result = net.forward("ear_ang45_3_sca20_r_tra20_r_e/out/MatMul");
    result *= 48;
    result += 48;

    ldmk.clear();
    for (int i = 0; i < 55; i++)
    ldmk.push_back(Point2d(result.at<float>(0, i * 2),
            result.at<float>(0, i * 2 + 1)));
}

void detectLandmark(vector<vector<Mat>> processedROI,
vector<vector<vector<Point2d>>> &landmarks,
vector<string> imageNames)
{
    for (int i = 0; i < processedROI.size(); i++)
    {
        printProgress(i, processedROI.size());

        auto images = processedROI[i];
        auto imageName = imageNames[i];

        vector<vector<Point2d>> ldmks;
        for (int j = 0; j < images.size(); j++)
        {
            auto imageROI = images[j];
            vector<Point2d> ldmk;
            _detectLandmarks(imageROI, ldmk);
            ldmks.push_back(ldmk);

            Mat outImage;
            drawLandmarks(imageROI, ldmk, outImage);
            // Exporting landmark image
            String id = to_string(i) + "_" + to_string(j);
            writeToFile(imageName, LANDMARK_PATH, outImage, id);
        }
        landmarks.push_back(ldmks);
    }
}*/

fun reduceDataSparsity(points: ArrayList<Point>, outPoints: ArrayList<Point>,
                       k: Int)
{
    var centroid = computeCentroid(points)
    // vector<Point2d> _points(points);
    // points.clear();
    outPoints.clear()

    var distances = arrayListOf<Double>()
    for (p in points) {
        // TODO: check these calculations
        var dx = (p.x - centroid.x).pow(2)
        var dy = (p.y - centroid.y).pow(2)
        distances.add(sqrt(dx + dy))
    }
    var mean = distances.average()
    var std = 0.0
    for (v in distances) std += (v - mean) * (v - mean)
    std = sqrt(std * (1.0f / distances.size))


    for (i in 0.until(points.size))
    {
        if (distances[i] < k * std)
            outPoints.add(points[i]);
    }
    if (outPoints.size < 2) {
        outPoints.clear()
        points.forEach { outPoints.add(it) }
    }
}

fun extractFeatures(images: ArrayList<ArrayList<Mat?>>,
                    landmarks: ArrayList<ArrayList<ArrayList<Point>>>)
{
    var keypoints = arrayListOf<MatOfKeyPoint>()
    var _t = arrayListOf<Mat?>()
    extractFeatures(images[0], keypoints, _t);

    // Translate KeyPoint -> Point2d
    for ( image in keypoints)
    {
        var imgLdmks = arrayListOf<ArrayList<Point>>()



        for (kpts in 0.until(image.cols())) {
            var ldmks = arrayListOf<Point>()
            for (k in 0.until(image.rows())) {
                ldmks.add(Point(image.get(k,kpts)[0], image.get(k,kpts)[1]))
            }

            imgLdmks.add(ldmks)
        }
        landmarks.add(imgLdmks);
    }

    var outLandmarks = arrayListOf<ArrayList<ArrayList<Point>>>()
    for ( image in landmarks)
    {
        var outLdmks = arrayListOf<ArrayList<Point>>()
        for (ldmk in image)
        {
            var outLdmk = arrayListOf<Point>()
            reduceDataSparsity(ldmk, outLdmk, 3);
            outLdmks.add(outLdmk)
        }
        outLandmarks.add(outLdmks)
    }


        var image = images[0]
        for (j in 0.until(image.size))
        {
            var outImage = Mat()

            landmarks[0][j] = outLandmarks[0][j];
        }

}




fun extractFeatures(images: ArrayList<ArrayList<Mat?>>, descriptors: ArrayList<Mat?>,
                     edgeThreshold: Int = 10,
                    mask: Mat = Mat())
{
    var keypoints = arrayListOf<MatOfKeyPoint>()
    var tmpDescriptors = ArrayList<Mat?>()
    extractFeatures(images[0], keypoints, tmpDescriptors);

    descriptors.clear();

    tmpDescriptors.forEach { descriptors.add(it) }

}




fun extractFeatures(images: ArrayList<Mat?>,
                    keypoints: ArrayList<MatOfKeyPoint>,
                    descriptors: ArrayList<Mat?>, edgeThreshold: Int = 10,
                    mask: Mat = Mat())
{
    var detector = create(500, 1.2f, 8, edgeThreshold);

    for (i in 0.until(images.size))
    {
        var image = images[i];
        var imageDescriptors = arrayListOf<Mat>()
        var imageKeypoints = arrayListOf<MatOfKeyPoint>()


            var ROIkeypoints = MatOfKeyPoint()
            var roiDescriptors = Mat()

            detector.detectAndCompute(image, mask, ROIkeypoints, roiDescriptors);
            //imageDescriptors.add(roiDescriptors);
            //imageKeypoints.add(ROIkeypoints);

        descriptors.add(roiDescriptors);
        keypoints.add(ROIkeypoints);
    }
}

fun computeSimilarity(queryDescriptors: Mat, objectDescriptors: Mat,
normType: Int = NORM_HAMMING, ratio: Double = 0.75, crossCheck: Boolean = false): Double
{
    var matches = arrayListOf<MatOfDMatch>()
    var matcher = BFMatcher.create(normType, crossCheck);
    matcher.knnMatch(queryDescriptors, objectDescriptors, matches, 2);

    var goodMatches = ArrayList<DMatch>()
    for (match in matches)
    {
        if (match.toList()[0].distance < ratio * match.toList()[1].distance)
        {
            goodMatches.add(match.toList()[0])
        }
    }
    return goodMatches.size.toDouble() /  matches.size.toDouble()
}
